src/workload.rs

```diff
@@ -118,51 +181,40 @@ impl fmt::Display for Upstream {
     }
 }
 
-fn byte_to_ip(b: bytes::Bytes) -> Result<Option<IpAddr>, WorkloadError> {
-    Ok(match b.len() {
-        0 => None,
-        4 => {
-            let v: [u8; 4] = (&*b).try_into().expect("size already proven");
-            Some(IpAddr::V4(Ipv4Addr::from(v)))
-        }
-        16 => {
-            let v: [u8; 16] = (&*b).try_into().expect("size already proven");
-            Some(IpAddr::V6(Ipv6Addr::from(v)))
-        }
-        n => return Err(WorkloadError::ByteAddressParse(n)),
-    })
+fn byte_to_ip(b: &bytes::Bytes) -> Result<Option<IpAddr>, WorkloadError> {
+    let s = std::str::from_utf8(b).expect("error converting bytes to string");
+    if s.is_empty() {
+        return Ok(None);
+    }
+    match IpAddr::from_str(s) {
+        Ok(ip) => Ok(Some(ip)),
+        Err(_) => Err(WorkloadError::ByteAddressParse(s.len())),
+    }
 }
 
 impl TryFrom<&XdsWorkload> for Workload {
     type Error = WorkloadError;
     fn try_from(resource: &XdsWorkload) -> Result<Self, Self::Error> {
-        let waypoint = byte_to_ip(resource.waypoint_address.clone())?;
-        let address =
-            byte_to_ip(resource.address.clone())?.ok_or(WorkloadError::ByteAddressParse(0))?;
-        // TODO can we borrow instead of clone everywhere?
-        Ok(Workload {
-            workload_ip: address,
-            waypoint_address: waypoint,
-            gateway_ip: None,
-
-            protocol: Protocol::try_from(xds::istio::workload::Protocol::from_i32(
-                resource.protocol,
-            ))?,
-
-            name: resource.name.clone(),
-            namespace: resource.namespace.clone(),
-            service_account: {
-                let result = resource.service_account.clone();
-                if result.is_empty() {
-                    "default".into()
-                } else {
-                    result
-                }
-            },
-            node: resource.node.clone(),
-
-            native_hbone: resource.native_hbone,
-        })
+        let resource = resource.to_owned();
+        let workload_ip =
+            byte_to_ip(&resource.address)?.ok_or(WorkloadError::ByteAddressParse(0))?;
+        let waypoint_ip = byte_to_ip(&resource.waypoint_address)?;
+        let protocol =
+            Protocol::try_from(xds::istio::workload::Protocol::from_i32(resource.protocol))?;
+        let service_account =
+            Option::from(resource.service_account).unwrap_or(String::from("default"));
+        let workload = WorkloadBuilder::new(
+            resource.name,
+            resource.namespace,
+            resource.native_hbone,
+            resource.node,
+            protocol,
+            service_account,
+            workload_ip,
+        )
+        .waypoint_address(waypoint_ip)
```

Comment:
Can you help my understand why we want to some fields in the `new` but this one as a function? Is it because its an Option? It seems like we are always setting it though.

Do we lose some type safety by switching from named struct to unnamed param list? For example, name and namespace could accidentally be swapped and not noticed?

Reply:
Generally, you want required fields in the `new` and optional fields as a function.

The unnamed param list is just a shortcut if you keep the param names the same as the field names.

Reply:
But we seem to have lost type safety (I can trivially cause a bug by swapping `name` and `namespace` - impossible with the struct shortcut) and readability, while making it more verbose .

Its not clear to me what value we are getting out of just using the struct directly?

Reply:
TBH this is my first Rust project... so while I have an OK handle on how to get things to compile at this point, my knowledge of best practices is pretty slim - very open to expert opinions here. I like the builder abstracting logic like that, but don't love that its easy to miss required fields. It does at least seem like we only actually make a Workload in 1 single place, so any minor deficiencies aren't a huge issue and I am ok with whatever is preferred. 

At the very least abstracting out some of the conversion logic could make sense

Reply:
Okay, I'll revert the builder stuff then.